drivers/char/tpm/tpm_ftpm_optee.c

```diff
@@ -0,0 +1,304 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ *
+ * Author: Jiri Appl <jiria@microsoft.com>
+ *
+ * OpTEE fTPM TA TPM driver
+ * 
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation, version 2 of the
+ * License.
+ *
+ */
+
+#include <linux/tee_drv.h>
+
+#include "tpm.h"
+
+#define TEEC_CONFIG_PAYLOAD_REF_COUNT 4
+#define TEEC_LOGIN_PUBLIC             0x00000000
+#define TEEC_SUCCESS                  0x00000000
+
+#define FTPM_OPTEE_TA_SUBMIT_COMMAND  0
+
+struct tpm_tee_context 
+{
+	struct tpm_chip *chip;
+
+	size_t response_length;
+	struct mutex response_buffer_lock;
+	u8 response_buffer[TPM_BUFSIZE];
+
+	struct tee_context *ftpm_optee_context;
+
+	bool optee_session_opened;
+	u32 optee_session_id;
+
+	bool chip_registered;
+};
+
+static struct tpm_tee_context tpm_tee_context = { 0 };
+
+static int ftpm_optee_tpm_op_recv(struct tpm_chip *chip, u8 *buf, size_t len)
+{
+	size_t response_length;
+	int rc;
+
+	mutex_lock(&tpm_tee_context.response_buffer_lock);
+
+	response_length = tpm_tee_context.response_length;
+	if (len < response_length) 
+	{
+		rc = -EIO;
+		goto cleanup;
+	}
+
+	memcpy(buf, tpm_tee_context.response_buffer, response_length);
```

Comment:
Is `buf` guaranteed to always be large enough to hold `tpm_tee_context.response_buffer`?

Reply:
Oh, never mind :) .